src/lightning/data/streaming/dataset.py

```diff
@@ -11,36 +11,135 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
-from typing import Any, Literal, Optional, Union
+from typing import Any, List, Literal, Optional, Union
 
-from torch.utils.data import Dataset
+import numpy as np
+from torch.utils.data import IterableDataset
 
+from lightning.data.datasets.env import _DistributedEnv, _WorkerEnv
 from lightning.data.streaming import Cache
+from lightning.data.streaming.item_loader import BaseItemLoader
+from lightning.data.streaming.sampler import ChunkedIndex
 
 
-class StreamingDataset(Dataset):
+class StreamingDataset(IterableDataset):
     """The streaming dataset can be used once your data have been optimised using the DatasetOptimiser class."""
 
     def __init__(
-        self, name: str, version: Optional[Union[int, Literal["latest"]]] = "latest", cache_dir: Optional[str] = None
+        self,
+        name: str,
+        version: Optional[Union[int, Literal["latest"]]] = "latest",
+        cache_dir: Optional[str] = None,
+        item_loader: Optional[BaseItemLoader] = None,
+        shuffle: bool = True,
+        seed: int = 42,
     ) -> None:
         """The streaming dataset can be used once your data have been optimised using the DatasetOptimiser class.
 
         Arguments:
             name: The name of the optimised dataset.
             version: The version of the dataset to use.
             cache_dir: The cache dir where the data would be stored.
+            item_loader: The logic to load an item from a chunk.
+            shuffle: Whether to shuffle the data.
+            seed: Random seed for shuffling.
 
         """
         super().__init__()
-        self.cache = Cache(name=name, version=version, cache_dir=cache_dir)
+        self.cache = Cache(name=name, version=version, cache_dir=cache_dir, item_loader=item_loader, chunk_bytes=1)
+
+        self.cache._reader._try_load_config()
+
+        if not self.cache.filled:
+            raise ValueError(f"The provided dataset `{name}` isn't filled up.")
+
+        self.shuffle = shuffle
+        self.distributed_env = _DistributedEnv.detect()
+        self.worker_env: Optional[_WorkerEnv] = None
+
+        chunk_intervals = self.cache.get_chunk_interval()
+        self.L = sum([(interval[-1] - interval[0]) for interval in chunk_intervals])
+
+        self.worker_chunks: List[int] = []
+        self.worker_intervals: List[List[int]] = []
+        self.current_indexes: List[int] = []
+        self.chunk_index = 0
+        self.index = 0
+        self.has_triggered_download = False
+        self.min_items_per_replica: Optional[int] = None
+        self.seed = seed
+        self.num_iter = 0
+        self.random_state = None
 
     def __len__(self) -> int:
-        return len(self.cache)
+        return self.L
+
+    def __iter__(self) -> "StreamingDataset":
```

Comment:
The lines below are important to cover with a test I think

Reply:
Yes, this is coming. This isn't supported yet for distributed training.